lib/spack/spack/database.py

```diff
@@ -1213,7 +1209,7 @@ def _remove(self, spec):
 
         # This install prefix is now free for other specs to use, even if the
         # spec is only marked uninstalled.
-        if not rec.spec.external:
+        if not rec.spec.external and rec.installed:
```

Comment:
This was also a bug: only part of the invariant was checked.

It was only hit thanks to the faulty test described in https://github.com/spack/spack/issues/28779: this test added an entry to the database, but it was not considered installed on the filesystem (spec passed to database did not correspond to spec on filesystem). So it ends up as registering an entry in the database with `rec.installed==False`, `rec.spec.external==False` and `rec.ref_count = 0`...

An actual test could be: a package that somehow destructs its own install prefix after installation and before being registered in the database. Although it's quite edge-case'y and unclear what expected behavior should be.

Reply:
This is now covered by a test.